R/eval-relocate.R

```diff
@@ -0,0 +1,163 @@
+#' Evaluate an expression to relocate variables
+#'
+#' @description
+#' `eval_relocate()` is a variant of [eval_select()] that moves a selection to
+#' a new location. Either `before` or `after` can be provided to specify where
+#' to move the selection to. This powers `dplyr::relocate()`.
+#'
+#' @inheritParams eval_select
+#'
+#' @param before,after Defused R code describing a selection according to the
+#'   tidyselect syntax. The selection represents the destination of the
+#'   selection provided through `expr`. Supplying neither of these will move the
+#'   selection to the left-hand side. Supplying both of these is an error.
+#'
+#' @param before_arg,after_arg Argument names for `before` and `after`. These
+#'   are used in error messages.
+#'
+#' @return
+#' A named vector of numeric locations with length equal to `length(data)`.
+#' Each position in `data` will be represented exactly once.
+#'
+#' The names are normally the same as in the input data, except when the user
+#' supplied named selections with `c()`. In the latter case, the names reflect
+#' the new names chosen by the user.
+#'
+#' @export
+#' @examples
+#' library(rlang)
+#'
+#' # Interpret defused code as a request to relocate
+#' x <- expr(c(mpg, disp))
+#' after <- expr(wt)
+#' eval_relocate(x, mtcars, after = after)
+#'
+#' # Supplying neither `before` nor `after` will move the selection to the
+#' # left-hand side
+#' eval_relocate(x, mtcars)
+#'
+#' # Within a function, use `enquo()` to defuse a single argument.
+#' # Note that `before` and `after` must also be defused with `enquo()`.
+#' my_relocator <- function(x, expr, before = NULL, after = NULL) {
+#'   eval_relocate(enquo(expr), x, before = enquo(before), after = enquo(after))
+#' }
+#'
+#' my_relocator(mtcars, vs, before = hp)
+#'
+#' # Here is an example of using `eval_relocate()` to implement `relocate()`.
+#' # Note that the dots are passed on as a defused call to `c(...)`.
+#' relocate <- function(.x, ..., .before = NULL, .after = NULL) {
+#'   pos <- eval_relocate(
+#'     expr(c(...)),
+#'     .x,
+#'     before = enquo(.before),
+#'     after = enquo(.after)
+#'   )
+#'   set_names(.x[pos], names(pos))
+#' }
+#'
+#' relocate(mtcars, vs, .before = hp)
+#' relocate(mtcars, starts_with("d"), .after = last_col())
+eval_relocate <- function(expr,
+                          data,
+                          ...,
+                          before = NULL,
+                          after = NULL,
+                          strict = TRUE,
+                          name_spec = NULL,
+                          allow_rename = TRUE,
+                          allow_empty = TRUE,
+                          before_arg = "before",
+                          after_arg = "after",
+                          env = caller_env(),
```

Comment:
I put `env` here, which is a little different from `eval_select()`. It felt better here since it is an optional argument? But I can move it after `data` if you'd like